azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_azurecr.go

```diff
@@ -0,0 +1,77 @@
+package serviceendpoint
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+	"github.com/microsoft/azure-devops-go-api/azuredevops/serviceendpoint"
+	"github.com/terraform-providers/terraform-provider-azuredevops/azuredevops/internal/utils/converter"
+)
+
+// ResourceServiceEndpointACR schema and implementation for ACR service endpoint resource
+func ResourceServiceEndpointAzureCR() *schema.Resource {
+	r := genBaseServiceEndpointResource(flattenServiceEndpointAzureCR, expandServiceEndpointAzureCR)
+	makeUnprotectedSchema(r, "azurecr_spn_tenantid", "ACR_TENANT_ID", "The service principal tenant id which should be used.")
+	makeUnprotectedSchema(r, "azurecr_subscription_id", "ACR_SUBSCRIPTION_ID", "The Azure subscription Id which should be used.")
+	makeUnprotectedSchema(r, "azurecr_subscription_name", "ACR_SUBSCRIPTION_NAME", "The Azure subscription name which should be used.")
+
+	r.Schema["resource_group"] = &schema.Schema{
+		Type:        schema.TypeString,
+		Required:    true,
+		Description: "Scope Resource Group",
+	}
+
+	r.Schema["azurecr_name"] = &schema.Schema{
+		Type:        schema.TypeString,
+		Required:    true,
+		DefaultFunc: schema.EnvDefaultFunc("AZDO_AZURECR_SERVICE_CONNECTION_REGISTRY", nil),
+		Description: "The AzureContainerRegistry registry which should be used.",
+	}
+
+	return r
+}
+
+// Convert internal Terraform data structure to an AzDO data structure
+func expandServiceEndpointAzureCR(d *schema.ResourceData) (*serviceendpoint.ServiceEndpoint, *string, error) {
+	serviceEndpoint, projectID := doBaseExpansion(d)
+	scope := fmt.Sprintf(
+		"/subscriptions/%s/resourceGroups/%s/providers/Microsoft.ContainerRegistry/registries/%s",
+		d.Get("azurecr_subscription_id"), d.Get("resource_group"), d.Get("azurecr_name"),
+	)
+	login_server := fmt.Sprintf("%s.azurecr.io", d.Get("azurecr_name"))
+	serviceEndpoint.Authorization = &serviceendpoint.EndpointAuthorization{
+		Parameters: &map[string]string{
+			"authenticationType": "spnKey",
+			"tenantId":           d.Get("azurecr_spn_tenantid").(string),
+			"loginServer":        login_server,
+			"scope":              scope,
+		},
+		Scheme: converter.String("ServicePrincipal"),
+	}
+	serviceEndpoint.Data = &map[string]string{
+		"registryId":       scope,
+		"subscriptionId":   d.Get("azurecr_subscription_id").(string),
```

Comment:
Extract `d.Get("azurecr_subscription_id")` into a variable since it's used in multiple places